ext/nnstreamer/tensor_decoder/tensordec-imagesegment.c

```diff
@@ -0,0 +1,287 @@
+ * 
+ * [DeeplabV3 model]
+ * Just one tensor with [21(#labels):width:height:1], float32
+ * Probability that each pixel is assumed to be labeled object.
+ */
+static GstCaps *
+is_getOutCaps (void **pdata, const GstTensorsConfig * config)
+{
+  image_segments *idata = *pdata;
+  gint fn, fd;
+  GstCaps *caps;
+  char *str;
+
+  if (idata->mode == TFLITE_DEEPLAB) {
+    g_return_val_if_fail (config != NULL, NULL);
+    GST_INFO ("Num Tensors = %d", config->info.num_tensors);
+    g_return_val_if_fail (config->info.num_tensors >= 1, NULL);
+
+    if (idata->width == 0 || idata->height == 0) {
+      idata->width = config->info.info[0].dimension[1];
+      idata->height = config->info.info[0].dimension[2];
+      _init_modes (idata);
+    }
+  }
+
+  str = g_strdup_printf ("video/x-raw, format = RGBA, "
+      "width = %u, height = %u", idata->width, idata->height);
+  caps = gst_caps_from_string (str);
+  fn = config->rate_n; /** @todo Verify if this rate is ok */
+  fd = config->rate_d;
+
+  if (fn >= 0 && fd > 0) {
+    gst_caps_set_simple (caps, "framerate", GST_TYPE_FRACTION, fn, fd, NULL);
+  }
+  g_free (str);
+
+  return gst_caps_simplify (caps);
+}
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+static size_t
+is_getTransformSize (void **pdata, const GstTensorsConfig * config,
+    GstCaps * caps, size_t size, GstCaps * othercaps, GstPadDirection direction)
+{
+  return 0;
+  /** @todo Use appropriate values */
+}
+
+
+/** @brief Set color according to each pixel's max probability  */
+static void
+set_color_according_to_label (image_segments * idata, GstMapInfo * out_info)
+{
+
+}
+
+
+/** @brief Set label index according to each pixel's label probabilities */
+static void
+set_label_index (image_segments * idata, void *data)
+{
+
+}
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+static GstFlowReturn
+is_decode (void **pdata, const GstTensorsConfig * config,
+    const GstTensorMemory * input, GstBuffer * outbuf)
+{
+  image_segments *idata = *pdata;
+  const size_t size = idata->width * idata->height * RGBA_CHANNEL;
+  GstMapInfo out_info;
+  GstMemory *out_mem;
+
+  g_assert (outbuf);
+  if (gst_buffer_get_size (outbuf) == 0) {
+    out_mem = gst_allocator_alloc (NULL, size, NULL);
+  } else {
+    if (gst_buffer_get_size (outbuf) < size) {
+      gst_buffer_set_size (outbuf, size);
+    }
+    out_mem = gst_buffer_get_all_memory (outbuf);
+  }
+  g_assert (gst_memory_map (out_mem, &out_info, GST_MAP_WRITE));
+
+  memset (out_info.data, 0, size);
+
+  if (idata->mode == TFLITE_DEEPLAB) {
+    g_assert (config->info.info[0].type == _NNS_FLOAT32);
+    g_assert (config->info.info[0].dimension[0] == TFLITE_DEEPLAB_TOTAL_LABELS);
+    set_label_index (idata, input->data);
+  }
+
+  set_color_according_to_label (idata, &out_info);
+
+  gst_memory_unmap (out_mem, &out_info);
+
+  if (gst_buffer_get_size (outbuf) == 0)
+    gst_buffer_append_memory (outbuf, out_mem);
+
+  return GST_FLOW_OK;
```

Comment:
Q: Does this function always return GST_FLOW_OK unless it dies with assertion?

Reply:
Yes